routing/result_interpretation.go

```diff
@@ -394,13 +394,108 @@ func (i *interpretedResult) processPaymentOutcomeIntermediate(
 	case *lnwire.FailExpiryTooSoon:
 		reportAll()
 
+	// We only expect to get FailInvalidBlinding from an introduction node
+	// in a blinded route. The introduction node in a blinded route is
+	// always responsible for reporting errors for the blinded portion of
+	// the route (to protect the privacy of the members of the route), so
+	// we need to be careful not to unfairly "shoot the messenger".
+	//
+	// The introduction node has no incentive to falsely report errors to
+	// sabotage the blinded route because:
+	//   1. Its ability to route this payment is strictly tied to the
+	//      blinded route.
+	//   2. The pubkeys in the blinded route are ephemeral, so doing so
+	//      will have no impact on the nodes beyond the individual payment.
+	//
+	// There are a few cases where we could unexpectedly receive this error:
+	// 1. Outside of a blinded route: erring node is not spec compliant.
+	// 2. Before the introduction point: erring node is not spec compliant.
+	// 3. After the introduction point: introduction node is not spec
+	//    compliant.
+	case *lnwire.FailInvalidBlinding:
+		introIdx, isBlinded := introductionPointIndex(route)
+
+		switch {
+		// Deal with cases where a node has incorrectly returned a
+		// blinding error:
+		// 1. A node before the introduction point returned it.
+		// 2. A node in a non-blinded route returned it.
+		case errorSourceIdx < introIdx || !isBlinded:
+			reportNode()
+
+		// If the error is after the introduction point, then the
+		// introduction node hasn't done its job of obfuscating the
+		// error. We fail the introduction node for not obeying the
+		// specification, and award success to any nodes before it
+		// that successfully forwarded the payment.
+		case errorSourceIdx > introIdx:
+			i.failNode(route, introIdx)
+
+			// Other preceding channels in the route forwarded
+			// correctly.
+			if introIdx > 1 {
+				i.successPairRange(route, 0, introIdx-2)
+			}
+
+		// Otherwise, the error was at the introduction node as we
+		// expected and we fail the final hop of the blinded route.
+		// We do this to minimize the amount of records that we store
+		// for the ephemeral blinded route, while still allowing
+		// retires.
+		default:
+			// All nodes up until the introduction node forwarded
+			// correctly.
+			if introIdx > 1 {
+				i.successPairRange(route, 0, introIdx-2)
```

Comment:
Hm, but I think we want to record that we were able to reach the introduction node? This is similar to a temporary channel failure with `reportOutgoingBalance`, where we mark all the route to the erring node successful, but the next hop is marked failed (in the blinded case the very last hop). In the tests it is typically the case that all the hops up to the erring node are marked successful, or do you have a specific case in mind that behaves differently?

Reply:
> do you have a specific case in mind that behaves differently

I was looking at `failNode`, where we don't grant a node the success on the incoming link (because we're going to fail its edges), but you're right that isn't the right comparison.